File: pca_test.py
#!/usr/bin/python
#
# Created by Albert Au Yeung (2010)
#
# An implementation of matrix factorization
#
import os
import math
try:
    import numpy
except:
    print "This implementation requires the numpy module."
    exit(0)

def load_reviews(path):
    maxUser = 0
    maxMovie = 0
    users = []

    for fileName in os.listdir(path):
        if os.path.isdir(path + fileName):
            continue

        with open(path + fileName, 'r') as f:
            movieId = int(f.readline()[:-2]) # Remove colon
            if movieId > maxMovie:
                maxMovie = movieId

            for line in f:
                splitReview = line.split(',')
                userId = int(splitReview[0])
                users.append(userId)

                if userId > maxUser:
                    maxUser = userId
    
    matrix = []
    #for i in range(maxUser):
    #    row = []
    #    for j in range(maxMovie):
    #        row.append(0.0)
    #    matrix.append(row)

    s = (maxUser, maxMovie)
    matrix = numpy.zeros(s)

    for fileName in os.listdir(path):
        if os.path.isdir(path + fileName):
            continue

        with open(path + fileName, 'r') as f:
            movieId = int(f.readline()[:-2]) # Remove colon
            for line in f:
                splitReview = line.split(',')
                userId = int(splitReview[0])
                rating = float(splitReview[1])
                
                matrix[userId - 1][movieId - 1] = rating
    return matrix, users 

def load_validation(path):
    reviews = []
    for fileName in os.listdir(path):
        if os.path.isdir(path + fileName):
            continue

        with open(path + fileName, 'r') as f:
            movieId = int(f.readline()[:-2]) # Remove colon
            for line in f:
                splitReview = line.split(',')
                review = {}
                review["movie"] = movieId
                review["user"] = int(splitReview[0])
                review["rating"] = int(splitReview[1])
                reviews.append(review)
    return reviews



###############################################################################

"""
@INPUT:
    R     : a matrix to be factorized, dimension N x M
    P     : an initial matrix of dimension N x K
    Q     : an initial matrix of dimension M x K
    K     : the number of latent features
    steps : the maximum number of steps to perform the optimisation
    alpha : the learning rate
    beta  : the regularization parameter
@OUTPUT:
    the final matrices P and Q
"""
def matrix_factorization(R, P, Q, K, steps=20, alpha=0.0002, beta=0.02):
    Q = Q.T
    for step in xrange(steps):
        print "Epoch: " + str(step + 1) + "/" + str(steps)
        for i in xrange(len(R)):
            for j in xrange(len(R[i])):
                if R[i][j] > 0:
                    eij = R[i][j] - numpy.dot(P[i,:],Q[:,j])
                    for k in xrange(K):
                        P[i][k] = P[i][k] + alpha * (2 * eij * Q[k][j] - beta * P[i][k])
                        Q[k][j] = Q[k][j] + alpha * (2 * eij * P[i][k] - beta * Q[k][j])
        print "Done prediction phase."
        eR = numpy.dot(P,Q)
        e = 0
        for i in xrange(len(R)):
            for j in xrange(len(R[i])):
                if R[i][j] > 0:
                    e = e + pow(R[i][j] - numpy.dot(P[i,:],Q[:,j]), 2)
                    for k in xrange(K):
                        e = e + (beta/2) * ( pow(P[i][k],2) + pow(Q[k][j],2) )
        print "Done training phase."
        if e < 0.001:
            break
    return P, Q.T

###############################################################################

if __name__ == "__main__":
    R = [
         [5,3,0,1],
         [4,0,0,1],
         [1,1,0,5],
         [1,0,0,4],
         [0,1,5,4],
        ]
    validation = load_validation("./dataset/training_set/training_set_test/validation/")

    R, users = load_reviews("./dataset/training_set/training_set_test/testing/")
    R = numpy.array(R)

    N = len(R)
    M = len(R[0])
    K = 900 

    P = numpy.random.rand(N,K)
    Q = numpy.random.rand(M,K)

    nP, nQ = matrix_factorization(R, P, Q, K)
    nR = numpy.dot(nP, nQ.T)
    print(nR)

    validationError = 0
    validationRMSE = 0
    validationCount = 0

    for review in validation:
        userId = review["user"]
        movieId = review["movie"]
        rating = review["rating"]

        if not userId in users:
            continue

        validationCount = validationCount + 1
        predictedRating = round(nR[userId-1][movieId-1])

        validationError = validationError + (abs(rating - predictedRating) / 5)
        validationRMSE = validationRMSE + pow(rating - predictedRating, 2)

    print "Validation Count: " + str(validationCount)
    print "Validation Accuracy: " + str((1 - (validationError / validationCount)) * 100)
    print "Validation RMSE: " + str(math.sqrt(validationRMSE / validationCount))

File: pca_test.lua
require 'nn' 
require 'cutorch'

MOVIE_ID_INDEX = 1
USER_ID_INDEX = 2
RATING_INDEX = 3

--Returns the names of all files in a given directory using find
function getFileNames(directory)
    local i, t, popen = 0, {}, io.popen
    for filename in popen('find "'..directory..'" -maxdepth 1 -type f -printf "%f\n"'):lines() do
        i = i + 1
        t[i] = filename
    end
    return t
end

--Standard string split method
function split(string, delim)
    if delim == nil then
        delim = "%s"
    end
    local t = {}
    local i = 1
    for str in string.gmatch(string, "([^"..delim.."]+)") do
        t[i] = str
        i = i + 1
    end

    return t
end

--Parse the raw review and put its information into a table
function createReview(rawReview)
    local splitReview = split(rawReview, ',')
    local review = {} 
    review[USER_ID_INDEX] = tonumber(splitReview[1])
    review[RATING_INDEX]  = tonumber(splitReview[2])
    --Note: movie ID is populated in loadDataset, so no, it wasn't forgotten!
    return review
end

--Parse a review string into a usable review table
function getReview(rawReview)
    if rawReview == nil then
        return nil
    end
    return createReview(rawReview)
end

--Load the movie rating information from the given directory into a table
--of movie tables which contain the movie ID as well as reviews associated
--with the movie
function loadDataset(directory)
    print("Loading dataset...")
    if directory == nil then return {} end

    local files = getFileNames(directory)
    local dataset = {}
    local users = {}
    local maxMovie = 0
    local maxUser = 0

    for i = 1, #files do
        if i % 100 == 0 then 
            collectgarbage()
        end
        local f = io.open(directory..files[i], 'r')
        local movieString = f:read()
        local movieId = tonumber(string.sub(movieString, 0, #movieString - 1))

        if movieId > maxMovie then
            maxMovie = movieId
        end

        while true do
            local review = getReview(f:read())
            if review == nil then break end
            
            if review[USER_ID_INDEX] > maxUser then
                maxUser = review[USER_ID_INDEX]
            end
        end
        f:close()
    end
   
    for i = 1, #files do
        if i % 100 == 0 then 
            collectgarbage()
        end
        local f = io.open(directory..files[i], 'r')
        local movieString = f:read()
        local movieId = tonumber(string.sub(movieString, 0, #movieString - 1))

        while true do
            local review = getReview(f:read())
            if review == nil then break end
            review[MOVIE_ID_INDEX] = movieId
            table.insert(users, review[USER_ID_INDEX])
            table.insert(dataset, review)
        end
        f:close()
    end
    ratingsMatrix = torch.Tensor(dataset)
    print("Dataset loaded.")
    return ratingsMatrix, users, maxUser, maxMovie 
end

function loadValidation(directory, users)
    print("Loading dataset...")
    if directory == nil then return {} end

    local files = getFileNames(directory)
    local dataset = {}

    for i = 1, #files do
        if i % 100 == 0 then 
            collectgarbage()
        end
        local f = io.open(directory..files[i], 'r')
        local movieString = f:read()
        local movieId = tonumber(string.sub(movieString, 0, #movieString - 1))

        while true do
            local review = getReview(f:read())
            if review == nil then break end
            review[MOVIE_ID_INDEX] = movieId

            if users[review[USER_ID_INDEX]] ~= nil then
                table.insert(dataset, review)
            end
        end
        f:close()
    end

    dataset = torch.Tensor(dataset)
    return dataset 
end

--[[
@INPUT:
    R     : a matrix to be factorized, dimension N x M
    P     : an initial matrix of dimension N x K
    Q     : an initial matrix of dimension M x K
    K     : the number of latent features
    steps : the maximum number of steps to perform the optimisation
    alpha : the learning rate
    beta  : the regularization parameter
@OUTPUT:
    the final matrices P and Q
]]--
function matrix_factorization(R, P, Q, K, steps, alpha, beta)
    Q = Q:t()
    for step = 1, steps do
        print(string.format("Epoch: %d/%d", step, steps))
        for i = 1, R:size(1) do
            
            local rating = R[i][RATING_INDEX]
            local movieId = R[i][MOVIE_ID_INDEX]
            local userId = R[i][USER_ID_INDEX]
            
            --TODO this could probably be vectorized a bit better
            local err = rating - torch.dot(P:select(1, userId), Q:select(2, movieId))
            for k = 1, K do
                P[userId][k] = P[userId][k] + alpha * (2 * err * Q[k][movieId] - beta * P[userId][k])
                Q[k][movieId] = Q[k][movieId] + alpha * (2 * err * P[userId][k] - beta * Q[k][movieId])
            end
        end

        print "Done prediction phase."
        
        e = 0
        for i = 1, R:size(1) do
            local rating = R[i][RATING_INDEX]
            local movieId = R[i][MOVIE_ID_INDEX]
            local userId = R[i][USER_ID_INDEX]
            e = e + torch.pow(rating - torch.dot(P:select(1, userId), Q:select(2, movieId)), 2)
            for k = 1, K do
                e = e + (beta/2) * ( torch.pow(P[userId][k],2) + torch.pow(Q[k][movieId],2) )
            end
        end
        
        print "Done training phase."
        if e < 0.001 then
            break
        end
    end
    return P, Q:t()
end


R, users, N, M = loadDataset("./dataset/training_set/training/")
R = R:cuda()

validation = loadValidation("./dataset/training_set/validation/", users):cuda()
K = 2 

P = torch.rand(N,K):cuda()
Q = torch.rand(M,K):cuda()

nP, nQ = matrix_factorization(R, P, Q, K, 1, 0.0002, 0.02)
nQ = nQ:t()
validationError = 0
validationRMSE = 0
validationCount = 0

for i = 1, validation:size(1) do
    review = validation[i]
    userId = review[USER_ID_INDEX]
    movieId = review[MOVIE_ID_INDEX]
    rating = review[RATING_INDEX]

    validationCount = validationCount + 1
    predictedRating = torch.round(torch.dot(nP:select(1, userId), nQ:select(2, movieId)))

    validationError = validationError + (torch.abs(rating - predictedRating) / 5)
    validationRMSE = validationRMSE + torch.pow(rating - predictedRating, 2)
end

print(string.format("Validation Count: %d", validationCount))
print(string.format("Validation Accuracy: %.4f", ((1 - (validationError / validationCount)) * 100)))
print(string.format("Validation RMSE: %.4f", math.sqrt(validationRMSE / validationCount)))
